From 13cab7308bf2ea3d70a774fbecd0f560178ddef1 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Wed, 3 Apr 2024 14:42:10 +0200
Subject: [PATCH 0745/1695] drivers: stm32_cpu_opp: prepare for scp-firmware
 CPU DVFS

Changes stm32_cpu_opp driver to provide to scp-firmware the platform
resources used for CPU operating point management that are the CPU
clock and the CPU voltage regulator. Also saves CPU OPP state during
PM suspend sequence so that it is restored in PM resume sequence.

Function setup_scmi_server_resources() sorts the operating point array
by increasing frequencies as expected by SCP-firmware DVFS module
configuration.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 3aa40a2594c81d803a7238e2467fc1bc23549745 drivers: stm32_cpu_opp: prepare for scp-firmware CPU DVFS]
Change-Id: I33ac235ce8046e50bed5906f747e67682dee93fd
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/371768
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/stm32_cpu_opp.c         | 332 ++++++++++++++++++++++++++-
 core/include/drivers/stm32_cpu_opp.h |  23 +-
 2 files changed, 345 insertions(+), 10 deletions(-)

diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index 68384ef60..ad9c85b6f 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -1,7 +1,9 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
- * Copyright (c) 2022, STMicroelectronics
+ * Copyright (c) 2022-2024, STMicroelectronics
  */
+
+#include <config.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
 #include <drivers/regulator.h>
@@ -17,34 +19,67 @@
 #include <kernel/panic.h>
 #include <kernel/pm.h>
 #include <libfdt.h>
+#ifdef CFG_SCMI_SCPFW
+#include <scmi_agent_configuration.h>
+#endif
 #include <stm32_util.h>
 #include <trace.h>
 #include <util.h>
 
+/*
+ * struct cpu_dvfs - CPU DVFS registered operating points
+ * @freq_khz: CPU frequency in kilohertz (kHz)
+ * @volt_uv: CPU voltage level in microvolts (uV)
+ */
 struct cpu_dvfs {
 	unsigned int freq_khz;
 	int volt_uv;
 };
 
+/*
+ * struct cpu_opp - CPU operating point
+ *
+ * @current_opp: Index of current CPU operating point in @dvfs array
+ * @opp_count: Number of cells of @dvfs
+ * @clock: CPU clock handle
+ * @regul: CPU regulator supply handle
+ * @dvfs: Arrays of the supported CPU operating points
+ * @scp_clock: Clock instance exposed to scp-firmware SCMI DVFS
+ * @scp_regulator: Regulator instance exposed to scp-firmware SCMI DVFS
+ * @scp_levels_desc: Description of voltage levels for scp-firmware SCMI DVFS
+ * @scp_cpu_opp_levels_uv: Array of voltage levels described by @scp_levels_desc
+ * @default_opp: Initial OPP state for scp-firmware
+ */
 struct cpu_opp {
 	unsigned int current_opp;
 	unsigned int opp_count;
 	struct clk *clock;
 	struct regulator *regul;
 	struct cpu_dvfs *dvfs;
+#ifdef CFG_SCPFW_MOD_DVFS
+	struct clk scp_clock;
+	struct regulator scp_regulator;
+	struct regulator_voltages_desc scp_levels_desc;
+	int *scp_cpu_opp_levels_uv;
+	unsigned int default_opp;
+#endif
 };
 
 static struct cpu_opp cpu_opp;
 
+#ifndef CFG_SCPFW_MOD_DVFS
 /* Mutex for protecting CPU OPP changes */
 static struct mutex cpu_opp_mu = MUTEX_INITIALIZER;
+#endif
 
 #define MPU_RAM_LOW_SPEED_THRESHOLD 1320000
 
+#ifndef CFG_SCPFW_MOD_DVFS
 size_t stm32_cpu_opp_count(void)
 {
 	return cpu_opp.opp_count;
 }
+#endif
 
 unsigned int stm32_cpu_opp_level(size_t opp_index)
 {
@@ -143,6 +178,251 @@ static TEE_Result set_voltage_then_clock(unsigned int opp)
 	return TEE_SUCCESS;
 }
 
+#ifdef CFG_SCPFW_MOD_DVFS
+/* Expose a CPU clock instance for scp-firmware DVFS module */
+static TEE_Result scp_set_cpu_rate(struct clk *clk __unused, unsigned long rate,
+				   unsigned long parent_rate __unused)
+{
+	unsigned int __maybe_unused khz = rate / 1000;
+
+#ifdef CFG_STM32MP15
+	res = stm32mp1_set_opp_khz(khz);
+	if (res)
+		return res;
+#else
+	if (rate != clk_get_rate(cpu_opp.clock))
+		return TEE_ERROR_GENERIC;
+#endif
+
+	return TEE_SUCCESS;
+}
+
+static unsigned long scp_read_cpu_rate(struct clk *clk __unused,
+				       unsigned long parent_rate)
+{
+	return parent_rate;
+}
+
+static TEE_Result scp_get_cpu_rates_array(struct clk *clk __unused,
+					  size_t start_index,
+					  unsigned long *rates,
+					  size_t *nb_elts)
+{
+	size_t rates_cells = *nb_elts;
+	size_t opp = 0;
+
+	if (start_index >= cpu_opp.opp_count)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	*nb_elts = cpu_opp.opp_count - start_index;
+
+	if (!rates || rates_cells < *nb_elts)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	for (opp = start_index; opp < cpu_opp.opp_count; opp++)
+		rates[opp] = (unsigned long)stm32_cpu_opp_level(opp) * 1000;
+
+	return TEE_SUCCESS;
+}
+
+static const struct clk_ops stm32_cpu_opp_clk_ops = {
+	.get_rate = scp_read_cpu_rate,
+	.set_rate = scp_set_cpu_rate,
+	.get_rates_array = scp_get_cpu_rates_array,
+};
+
+/*
+ * Expose a regulator for PSU (Power Supply Unit) instance used
+ * by scp-firmware DVFS module
+ */
+static TEE_Result scp_set_regu_state(struct regulator *regulator __unused,
+				     bool enabled __unused)
+{
+	return TEE_SUCCESS;
+}
+
+static TEE_Result scp_read_regu_state(struct regulator *regulator __unused,
+				      bool *enabled)
+{
+	*enabled = true;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result scp_set_regu_voltage(struct regulator *regulator __unused,
+				       int uv)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int cur_uv = 0;
+
+	cur_uv = regulator_get_voltage(cpu_opp.regul);
+	if (uv == cur_uv)
+		return TEE_SUCCESS;
+
+#ifdef CFG_STM32MP13
+	if (cur_uv > MPU_RAM_LOW_SPEED_THRESHOLD &&
+	    uv <= MPU_RAM_LOW_SPEED_THRESHOLD)
+		io_setbits32(stm32_pwr_base(), PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+	res = opp_set_voltage(cpu_opp.regul, uv);
+	if (res)
+		return res;
+
+#ifdef CFG_STM32MP13
+	if (cur_uv <= MPU_RAM_LOW_SPEED_THRESHOLD &&
+	    uv > MPU_RAM_LOW_SPEED_THRESHOLD)
+		io_clrbits32(stm32_pwr_base(), PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result scp_read_regu_voltage(struct regulator *regulator __unused,
+					int *uv)
+{
+	*uv = regulator_get_voltage(cpu_opp.regul);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result scp_regu_voltages(struct regulator *regulator __unused,
+				    struct regulator_voltages_desc **desc,
+				    const int **levels)
+{
+	*desc = &cpu_opp.scp_levels_desc;
+	*levels = cpu_opp.scp_cpu_opp_levels_uv;
+
+	return TEE_SUCCESS;
+}
+
+static const struct regulator_ops stm32_scp_cpu_opp_regu = {
+	.set_state = scp_set_regu_state,
+	.get_state = scp_read_regu_state,
+	.set_voltage = scp_set_regu_voltage,
+	.get_voltage = scp_read_regu_voltage,
+	.supported_voltages = scp_regu_voltages,
+};
+
+static int cmp_cpu_opp_by_freq(const void *a, const void *b)
+{
+	const struct cpu_dvfs *opp_a = a;
+	const struct cpu_dvfs *opp_b = b;
+
+	if (opp_a->freq_khz == opp_b->freq_khz)
+		return CMP_TRILEAN(opp_a->volt_uv, opp_b->volt_uv);
+	else
+		return CMP_TRILEAN(opp_a->freq_khz, opp_b->freq_khz);
+}
+
+static int min_cpu_voltage(struct cpu_dvfs *dvfs, size_t count)
+{
+	int min_mv = INT_MAX;
+	size_t n = 0;
+
+	for (n = 0; n < count; n++)
+		min_mv = MIN(min_mv, dvfs[n].volt_uv);
+
+	assert(min_mv < INT_MAX);
+	return min_mv;
+}
+
+static int max_cpu_voltage(struct cpu_dvfs *dvfs, size_t count)
+{
+	int max_mv = INT_MIN;
+	size_t n = 0;
+
+	for (n = 0; n < count; n++)
+		max_mv = MAX(max_mv, dvfs[n].volt_uv);
+
+	assert(max_mv > 0);
+	return max_mv;
+}
+
+static void setup_scp_server_resources(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int *dvfs_khz = NULL;
+	unsigned int *dvfs_mv = NULL;
+	size_t opp = 0;
+	struct cpu_dvfs *sorted_dvfs = NULL;
+
+	/*
+	 * Sort operating points by increasing frequencies as expected by
+	 * SCP-firmware DVFS module.
+	 */
+	sorted_dvfs = calloc(cpu_opp.opp_count, sizeof(*sorted_dvfs));
+	assert(sorted_dvfs);
+	memcpy(sorted_dvfs, cpu_opp.dvfs,
+	       cpu_opp.opp_count * sizeof(*sorted_dvfs));
+
+	qsort(sorted_dvfs, cpu_opp.opp_count, sizeof(*sorted_dvfs),
+	      cmp_cpu_opp_by_freq);
+
+	/* Setup  a clock for scp-firmare DVFS module clock instance */
+	cpu_opp.scp_clock = (struct clk){
+		.ops = &stm32_cpu_opp_clk_ops,
+		.name = "stm32-cpu-opp",
+		.rate = cpu_opp.dvfs[cpu_opp.current_opp].freq_khz * 1000,
+#ifndef CFG_STM32MP15
+		.flags = CLK_SET_RATE_PARENT,
+		.parent = cpu_opp.clock,
+#endif
+	};
+
+	res = clk_register(&cpu_opp.scp_clock);
+	if (res)
+		panic();
+
+	/* Setup a regulator for scp-firmare DVFS module PSU instance */
+	cpu_opp.scp_regulator = (struct regulator){
+		.ops = &stm32_scp_cpu_opp_regu,
+		.min_uv = min_cpu_voltage(sorted_dvfs, cpu_opp.opp_count),
+		.max_uv = max_cpu_voltage(sorted_dvfs, cpu_opp.opp_count),
+		.name = (char *)"stm32-cpu-opp",
+	};
+
+	cpu_opp.scp_levels_desc = (struct regulator_voltages_desc){
+		.type = VOLTAGE_TYPE_FULL_LIST,
+		.num_levels = cpu_opp.opp_count,
+	};
+
+	cpu_opp.scp_cpu_opp_levels_uv =
+		calloc(cpu_opp.opp_count,
+		       sizeof(*cpu_opp.scp_cpu_opp_levels_uv));
+
+	assert(cpu_opp.scp_cpu_opp_levels_uv);
+
+	for (opp = 0; opp < cpu_opp.opp_count; opp++)
+		cpu_opp.scp_cpu_opp_levels_uv[opp] = sorted_dvfs[opp].volt_uv;
+
+	/* Feed SCP-firmware with CPU DVFS configuration data */
+	dvfs_khz = calloc(cpu_opp.opp_count, sizeof(*dvfs_khz));
+	dvfs_mv = calloc(cpu_opp.opp_count, sizeof(*dvfs_mv));
+	assert(dvfs_khz && dvfs_mv);
+
+	for (opp = 0; opp < cpu_opp.opp_count; opp++) {
+		dvfs_khz[opp] = sorted_dvfs[opp].freq_khz;
+		dvfs_mv[opp] = sorted_dvfs[opp].volt_uv / U(1000);
+	}
+
+	free(sorted_dvfs);
+
+	res = scmi_scpfw_cfg_add_dvfs(0 /*agent*/, 0 /*channel*/, 0 /*domain*/,
+				      &cpu_opp.scp_regulator,
+				      &cpu_opp.scp_clock,
+				      dvfs_khz, dvfs_mv, cpu_opp.opp_count);
+	if (res)
+		panic();
+
+	free(dvfs_khz);
+	free(dvfs_mv);
+}
+#else /*CFG_SCPFW_MOD_DVFS*/
+static void setup_scp_server_resources(void)
+{
+}
+
 TEE_Result stm32_cpu_opp_set_level(unsigned int level)
 {
 	unsigned int current_level = 0;
@@ -190,18 +470,54 @@ TEE_Result stm32_cpu_opp_read_level(unsigned int *level)
 
 	return TEE_SUCCESS;
 }
+#endif /*CFG_SCPFW_MOD_DVFS*/
 
 #ifdef CFG_STM32MP13
 static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
 			     const struct pm_callback_handle *hdl __unused)
 {
 	unsigned long clk_cpu = 0;
-	unsigned int opp = cpu_opp.current_opp;
+	unsigned int opp = 0;
 
 	assert(op == PM_OP_SUSPEND || op == PM_OP_RESUME);
 
+#ifdef CFG_SCPFW_MOD_DVFS
+	if (op == PM_OP_SUSPEND) {
+		/*
+		 * When CFG_SCPFW_MOD_DVFS is enabled, save CPU OPP on suspend
+		 * for restoration at resmue. If CPU is not in an expected
+		 * OPP state, fallback to default OPP at PM resume.
+		 */
+		unsigned int cur_khz = 0;
+		int cur_uv = 0;
+
+		cur_khz = clk_get_rate(cpu_opp.clock) / 1000;
+
+		cur_uv = regulator_get_voltage(cpu_opp.regul);
+		if (!cur_uv)
+			panic();
+
+		/* Get current OPP or fallback to default */
+		for (opp = 0; opp < cpu_opp.opp_count; opp++)
+			if (cur_khz == cpu_opp.dvfs[opp].freq_khz  &&
+			    cur_uv == cpu_opp.dvfs[opp].volt_uv)
+				break;
+		if (opp >= cpu_opp.opp_count) {
+			EMSG("Unexpected OPP state, select default");
+			opp = cpu_opp.default_opp;
+		}
+
+		DMSG("Suspend to OPP %u", opp);
+		cpu_opp.current_opp = opp;
+
+		return TEE_SUCCESS;
+	}
+#endif
+
 	if (op == PM_OP_RESUME) {
-		DMSG("cpu opp resume opp to %u", opp);
+		opp = cpu_opp.current_opp;
+
+		DMSG("Resume to OPP %u", opp);
 
 		clk_cpu = clk_get_rate(cpu_opp.clock);
 		assert(clk_cpu);
@@ -313,6 +629,9 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 		return TEE_ERROR_GENERIC;
 
 	/* Select the max "st,opp-default" node as current OPP */
+#ifdef CFG_SCPFW_MOD_DVFS
+	cpu_opp.default_opp = cpu_opp.current_opp;
+#endif
 	clk_cpu = clk_get_rate(cpu_opp.clock);
 	assert(clk_cpu);
 	if (freq_khz_opp_def * 1000U > clk_cpu)
@@ -376,8 +695,13 @@ stm32_cpu_opp_init(const void *fdt, int node, const void *compat_data __unused)
 #endif
 
 	res = stm32_cpu_opp_get_dt_subnode(fdt, node);
+	if (res)
+		return res;
 
-	return res;
+	if (IS_ENABLED(CFG_SCPFW_MOD_DVFS))
+		setup_scp_server_resources();
+
+	return TEE_SUCCESS;
 }
 
 static const struct dt_device_match stm32_cpu_opp_match_table[] = {
diff --git a/core/include/drivers/stm32_cpu_opp.h b/core/include/drivers/stm32_cpu_opp.h
index 3f7d56d73..1736aa5ec 100644
--- a/core/include/drivers/stm32_cpu_opp.h
+++ b/core/include/drivers/stm32_cpu_opp.h
@@ -1,21 +1,32 @@
-/* SPDX-License-Identifier: BSD-3-Clause */
+/* SPDX-License-Identifier: BSD-2-Clause */
 /*
- * Copyright (c) 2022, STMicroelectronics
+ * Copyright (c) 2022-2024, STMicroelectronics
  */
 
-#ifndef __STM32_CPU_OPP__
-#define __STM32_CPU_OPP__
+#ifndef DRIVERS_STM32_CPU_OPP_H
+#define DRIVERS_STM32_CPU_OPP_H
 
+struct regulator;
+struct clk;
+
+#ifndef CFG_SCPFW_MOD_DVFS
 /* Return the number of CPU operating points */
 size_t stm32_cpu_opp_count(void);
+#endif
 
 /* Get level value identifying CPU operating point @opp_index */
 unsigned int stm32_cpu_opp_level(size_t opp_index);
 
+#ifdef CFG_SCPFW_MOD_DVFS
+/* Provide CPU clock and regulator reference for CPU DVFS with SCMI */
+void scmi_server_set_cpu_resources(struct regulator *regulator, struct clk *clk,
+				   unsigned int *dvfs_khz,
+				   unsigned int *dvfs_mv, size_t dvfs_count);
+#else
 /* Request to switch to CPU operating point related to @level */
 TEE_Result stm32_cpu_opp_set_level(unsigned int level);
 
 /* Get level related to current CPU operating point */
 TEE_Result stm32_cpu_opp_read_level(unsigned int *level);
-
-#endif /*__STM32_CPU_OPP__*/
+#endif /*CFG_SCPFW_MOD_DVFS*/
+#endif /*DRIVERS_STM32_CPU_OPP_H*/
-- 
2.47.2

